lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,276 @@
+  }
+
+  return indices;
+}
+
+void insertSubview(ArrayRef<int64_t> tensorShape, Type type, Type resultType,
+                   SmallVector<ReassociationIndices> reassociation,
+                   Value operand, ForallOp op, OpBuilder b,
+                   xsmm::BrgemmOp brgemmOp, int operandNumber) {
+
+  auto expandShape = b.create<memref::ExpandShapeOp>(
+      op.getLoc(),
+      MemRefType::get({tensorShape},
+                      dyn_cast<MemRefType>(type).getElementType()),
+      operand, reassociation);
+  expandShape.setStaticOutputShape(tensorShape);
+  b.setInsertionPoint(brgemmOp);
+  SmallVector<OpFoldResult> strides(tensorShape.size(), b.getIndexAttr(1)),
+      sizes, offsets;
+  size_t tileSize =
+      tensorShape.size() - dyn_cast<ShapedType>(resultType).getShape().size();
+
+  SmallVector<int64_t> tileSizes;
+  for (size_t i = 0; i < tensorShape.size(); i++) {
+    if (i < tileSize) {
+      int opnum = operandNumber;
+      if (opnum == 3) {
+        opnum = 1;
+      }
+      int inductionVarIndex = (opnum - 1) * tileSize + i;
+      offsets.push_back(op.getInductionVars()[inductionVarIndex]);
+      sizes.push_back(b.getIndexAttr(1));
+    } else {
+      sizes.push_back(b.getIndexAttr(tensorShape[i]));
+      tileSizes.push_back(tensorShape[i]);
+      offsets.push_back(b.getIndexAttr(0));
+    }
+  }
+
+  auto subviewType =
+      MemRefType::get({tileSizes}, dyn_cast<MemRefType>(type).getElementType());
+  auto [originalStride, originalOffset] =
+      getStridesAndOffset(dyn_cast<MemRefType>(subviewType));
+  subviewType = MemRefType::get(
+      {tileSizes}, dyn_cast<MemRefType>(subviewType).getElementType(),
+      StridedLayoutAttr::get(b.getContext(), ShapedType::kDynamic,
+                             originalStride));
+  auto subview = b.create<memref::SubViewOp>(
+      op.getLoc(), dyn_cast<MemRefType>(subviewType), expandShape.getResult(),
+      offsets, sizes, strides);
+  brgemmOp.getOperand(operandNumber).replaceAllUsesWith(subview);
+}
+
+FailureOr<ForallOp> insertParallelLoop(ForallOp op,
+                                       ArrayRef<unsigned> tileShapeM,
+                                       ArrayRef<unsigned> tileShapeN) {
+  xsmm::BrgemmOp brgemmOp = NULL;
+  OpBuilder b(op);
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++)
+    if (dyn_cast<xsmm::BrgemmOp>(oper)) {
+      brgemmOp = dyn_cast<xsmm::BrgemmOp>(oper);
+      break;
+    }
+  if (brgemmOp == NULL)
+    return failure();
+
+  int boundSize = tileShapeM.size() + tileShapeN.size();
+  auto mShape =
+      dyn_cast<ShapedType>(
+          brgemmOp.getOperand(1).getDefiningOp()->getOperand(0).getType())
+          .getShape();
+
+  // Validate the input tile sizes against the operand shapes
+  long multipleM = 1;
+  for (size_t i = 0; i < tileShapeM.size(); i++)
+    multipleM = multipleM * tileShapeM[i];
+
+  if (mShape[0] != multipleM)
+    return failure();
```

Comment:
nit: add some `notifyMatchFailure` message to make it easier to test and debug

I tried the pass on simple example and I have no clue why nothing happens.